### PR TITLE
WIP: Output data wells sans vectors

### PR DESCRIPTION
**Pull request by jokva**
A downstream change (and actual fix, this was always broken) for https://github.com/OPM/opm-output/pull/98

@totto82 Please help me test this.

### CODE REVIEW

**Comment by atgeirr** [comment]
I have to admit that I got a little confused about the role of this PR vs. OPM/opm-core#1075, but I think this just does the same for the derived well state classes (output, not input though) as that PR does for WellState. Correct?

**Comment by atgeirr** [comment]
> this was always broken

What was always broken?

**Comment by jokva** [comment]
Completions were never restored is the big one.

>

I have to admit that I got a little confused about the role of this PR vs. OPM/opm-core#1075, but I think this just does the same for the derived well state classes (output, not input though) as that PR does for WellState. Correct?

Yes.

**Comment by atgeirr** [comment]
What is the status of this chain of pull requests? Is the work sufficiently near completion that it can be reviewed and merged within a day or two?

**Comment by jokva** [comment]
It's complete save the perfrates restoration. Can't too easily hook into that since only the child class WellState is used, but the perfrates is in WellStateFullyImplicitBlackoil or whatever it's called and I seem to remember only the parent class is exposed. I leave it to you to figure out how to extract the perf rates and correctly populate the correct wellstate object - output's exposing it just fine.

**Comment by totto82** [comment]
I have started testing it. I will report back.

File: opm/autodiff/Compat.hpp [hunk 1]
```diff
@@ -169,14 +172,59 @@ inline void solutionToSim( const data::Solution& sol,
...
+            return p.second.pressure;
+        };
+
+        const auto perforation_reservoir_rate = []( const perfpair& p ) {
+            return p.second.reservoir_rate;
+        };
+
+        std::transform( well.completions.begin(),
+                        well.completions.end(),
+                        state.perfPress().begin() + well_index,
```
Comment (atgeirr) [inline opm/autodiff/Compat.hpp]: The index here is incorrect. It is the well index but should be the position of the first well perforation. This you can get from `wm.second[1]`.

File: opm/autodiff/Compat.hpp [hunk 2]
```diff
@@ -169,14 +172,59 @@ inline void solutionToSim( const data::Solution& sol,
...
+        state.bhp()[ well_index ] = well.bhp;
+        state.temperature()[ well_index ] = well.temperature;
+
+        const auto wellrate_index = well_index * np;
+        for( size_t i = 0; i < phs.size(); ++i ) {
+            assert( well.rates.has( phs[ i ] ) );
+            state.wellRates()[ wellrate_index + i ] = well.rates.get( phs[ i ] );
+        }
+
+        using perfpair = decltype( *well.completions.begin() );
```
Comment (atgeirr) [inline opm/autodiff/Compat.hpp]: Capitalize types.

File: opm/autodiff/Compat.hpp [hunk 3]
```diff
@@ -169,14 +172,59 @@ inline void solutionToSim( const data::Solution& sol,
...
+        };
+
+        std::transform( well.completions.begin(),
+                        well.completions.end(),
+                        state.perfPress().begin() + well_index,
+                        perforation_pressure );
+
+        std::transform( well.completions.begin(),
+                        well.completions.end(),
+                        state.perfPress().begin() + well_index,
```
Comment (atgeirr) [inline opm/autodiff/Compat.hpp]: You are writing to the wrong vector!

In addition the index is incorrect as above.

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 4]
```diff
@@ -196,9 +196,23 @@ namespace Opm
             data::Wells res = WellState::report(pu);
 
             const int nw = this->numWells();
-            // If there are now wells numPhases throws a floating point
-            // exception.
-            const int np = nw ? this->numPhases() : -1;
+            if( nw == 0 ) return res;
+            const int np = pu.num_phases;
+
+
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Huge gap here, intending to insert some code?

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 5]
```diff
@@ -196,9 +196,23 @@ namespace Opm
...
+                phs.at( pu.phase_pos[BlackoilPhases::Aqua] ) = rt::wat;
+            }
+
+            if( pu.phase_used[BlackoilPhases::Liquid] ) {
+                phs.at( pu.phase_pos[BlackoilPhases::Liquid] ) = rt::oil;
+            }
+
+            if( pu.phase_used[BlackoilPhases::Vapour] ) {
+                phs.at( pu.phase_pos[BlackoilPhases::Vapour] ) = rt::gas;
+            }
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: The part above is repeated from the other function, please refactor into own function.

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 6]
```diff
@@ -207,47 +221,17 @@ namespace Opm
...
-                }
-
-                if( pu.phase_used[BlackoilPhases::Liquid] ) {
-                    wellrates.set( rt::oil, wv[ wellrate_index + pu.phase_pos[BlackoilPhases::Liquid] ] );
-                }
-
-                if( pu.phase_used[BlackoilPhases::Vapour] ) {
-                    wellrates.set( rt::gas, wv[ wellrate_index + pu.phase_pos[BlackoilPhases::Vapour] ] );
+                for( auto& cpair : well.completions ) {
+                    const auto rates = this->perfPhaseRates().begin() + (np * w);
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Incorrect index. Here you must use `np * wt.second[1]` not `np * wt.second[0]` (as this is equivalent to) to get an iterator to the *start of* a well's perforation data. Then you must add `np * local_comp_index` to get the data for the correct index, where `local_comp_index` is the index of the completion within the set of completions for that single well (i.e. 0...(nwc-1), where nwc == `wt.second[2]` == the number of completions for the well in question).

**Comment by atgeirr** [comment]
There seems to be a missing feature (assuming all index bugs are fixed): calling wellsToState() will *not* restore all that was written in report() for a WellStateFullyImplicitBlackoil object. The simplest way would be add an overload that first calls the existing wellsToState(). (A more elegant solution might make all of these functions virtual methods of WellState, but that might hit other snags, have not evaluated that option carefully).

**Comment by atgeirr** [comment]
jenkins build this opm-output=98 opm-core=1075 run spe please

**Comment by atgeirr** [comment]
jenkins build this opm-output=98 opm-core=1075 run spe please

**Comment by atgeirr** [comment]
jenkins build this opm-output=98 opm-core=1075 run spe please

**Comment by atgeirr** [comment]
I do not understand why Jenkins fails on this one. It complains about an error related to a small opm-core API change that was merged this morning, perhaps rebasing all the PRs would help. In any case, I have tested this and believe it is fine now, so I'll merge.

In case I'm wrong, I'll fix it.

**Pull request merged**